### PR TITLE
[Safe-transaction-service] Overview and offchain signatures documentation

### PR DESCRIPTION
**Pull request by moisses89**
The idea of this PR is add more relevant documentation about safe transaction service based on users/developers questions. 
This PR is adding: 
- an overview of transaction service
- key features of transaction service
- off-chain signatures documentation
- messages feature
- transaction decoder feature

# Related issues
https://github.com/safe-global/safe-transaction-service/issues/1656

### CODE REVIEW

**Review by Uxio0 — APPROVED** [review]
Really nice work

File: safe-core-api/safe-transaction-service.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+- [Gunicorn](https://gunicorn.org/): it is a python WSGI HTTP Server.
+- [Celery](https://docs.celeryq.dev/en/stable/): It’s a task queue with focus on real-time processing, while also supporting task scheduling. Safe transaction service currently have a scheduler (to beat periodic tasks), a worker indexer to consume and execute indexing tasks and contracts worker mainly to get contracts metadata.
+- [RabbitMq](https://www.rabbitmq.com/): It is is a distributed message broker system used by celery to share the messages between scheduler, workers and django app.
+- [PostgreSql](https://www.postgresql.org/): it is an open source object-relational database system.
+- [Redis](https://redis.com/): it an open source, in-memory data structure store that can be used as a database, cache, message broker, and streaming engine. It is the cache of safe transaction service.
+- [safe-eth-py](https://github.com/safe-global/safe-eth-py) library to interact with Safe and blockchain.
+
+<figure><img src="../.gitbook/assets/transaction_service_architecture.png" width="100%" alt="" /></figure>
+
+## Offchain transaction signatures
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:26]:
```suggestion
## Offchain Transaction Signatures
```

File: safe-core-api/safe-transaction-service.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+The transaction service can collect off-chain transaction signatures, allowing the owners to share their signatures between them to reach the required threshold before execute a transaction and spending less gas than on-chain approvals.
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:27]:
```suggestion
The transaction service can collect offchain transaction signatures, allowing the owners to share their signatures to reach the required threshold before executing a transaction and spending less gas than onchain approvals.
```

File: safe-core-api/safe-transaction-service.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+The following endpoints let us to propose a transaction and collect every confirmation (off-chain signatures):
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:29]:
```suggestion
The following endpoints let us propose a transaction and collect every confirmation (offchain signatures):
```

File: safe-core-api/safe-transaction-service.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+- `POST /v1/safes/{address}/multisig-transactions/` create a new transaction, at least one signature is required
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:31]:
```suggestion
- `POST /v1/safes/{address}/multisig-transactions/`: Create a new transaction. At least one signature is required.
```

File: safe-core-api/safe-transaction-service.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+- `POST /v1/multisig-transactions/{safe_tx_hash}/confirmations/` add a new confirmation, the `safe_tx_hash` is needed
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:32]:
```suggestion
- `POST /v1/multisig-transactions/{safe_tx_hash}/confirmations/`: Add a new confirmation. The `safe_tx_hash` is needed.
```

File: safe-core-api/safe-transaction-service.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+- `GET /v1/multisig-transactions/{safe_tx_hash}/` return all the multisig-transaction information
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:33]:
```suggestion
- `GET /v1/multisig-transactions/{safe_tx_hash}/`: Return all the multisig transaction information.
```

File: safe-core-api/safe-transaction-service.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+- `GET /v1/multisig-transactions/{safe_tx_hash}/confirmations/` return the list of all confirmations to a multisig-transaction
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:34]:
```suggestion
- `GET /v1/multisig-transactions/{safe_tx_hash}/confirmations/`: Return the list of all confirmations to a multisig transaction.
```

File: safe-core-api/safe-transaction-service.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+The following sequence diagram shows an use case from a Safe shared by Alice and Bob where at least 1 confirmation for each one is required:
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:36]:
```suggestion
The following sequence diagram shows a use case for a Safe shared by Alice and Bob where at least one confirmation for each one is required:
```

File: safe-core-api/safe-transaction-service.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    A->>+SafeTransactionService: proposeTransaction POST /v1/safes/0x5Afe/multisig-transactions/
+    SafeTransactionService-->>-A: Http(201) {Created}
+    B->>+SafeTransactionService: GET /v1/safes/0x5Afe/multisig-transactions/?executed=false
+    SafeTransactionService-->>-B: Http(200) {..."safeTxHash":"0x5afe0001"}
+    B->>+SafeTransactionService: confirmTransaction POST /v1/multisig-transactions/0x5afe0001/confirmations/
+    SafeTransactionService->>-B: Http(201) {Created}
+```
+**What is the safe_tx_hash?**
+     
+`safe_tx_hash` is the unique identifier for a safe transaction and is calculated following [EIP-712](https://eips.ethereum.org/EIPS/eip-712) standard:
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:50]:
```suggestion
`safe_tx_hash` is the unique identifier for a Safe transaction and is calculated using the [EIP-712](https://eips.ethereum.org/EIPS/eip-712) standard:  
```

File: safe-core-api/safe-transaction-service.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    B->>+SafeTransactionService: GET /v1/safes/0x5Afe/multisig-transactions/?executed=false
+    SafeTransactionService-->>-B: Http(200) {..."safeTxHash":"0x5afe0001"}
+    B->>+SafeTransactionService: confirmTransaction POST /v1/multisig-transactions/0x5afe0001/confirmations/
+    SafeTransactionService->>-B: Http(201) {Created}
+```
+**What is the safe_tx_hash?**
+     
+`safe_tx_hash` is the unique identifier for a safe transaction and is calculated following [EIP-712](https://eips.ethereum.org/EIPS/eip-712) standard:  
+`keccak256(0x19 || 0x1 || domainSeparator || safeTxHashStruct)`  
+where `safeTxHashStruct` is the hashStruct of a safe transaction.
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:52]:
```suggestion
where `safeTxHashStruct` is the hashStruct of a Safe transaction.
```

File: safe-core-api/safe-transaction-service.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+eth_client = EthereumClient("https://rpc.goerli.eth.gateway.fm")
+safe_tx = SafeTx(eth_client, "0x4127839cdf4F73d9fC9a2C2861d8d1799e9DF40C", "0xc6b82bA149CFA113f8f48d5E3b1F78e933e16DfD",10000000000000000,"",0,0,0,0,"0x0000000000000000000000000000000000000000","0x0000000000000000000000000000000000000000", safe_nonce=206)
+print(safe_tx.safe_tx_hash.hex())
+```
+**Output**
+``` 
+0x34ae46cf7d884309a438a7e9a3161fa05dfc5068681ac3877a947971af845a18
+```
+
+## Offchain messages
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:67]:
```suggestion
## Offchain Messages 
```

File: safe-core-api/safe-transaction-service.md [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+safe_tx = SafeTx(eth_client, "0x4127839cdf4F73d9fC9a2C2861d8d1799e9DF40C", "0xc6b82bA149CFA113f8f48d5E3b1F78e933e16DfD",10000000000000000,"",0,0,0,0,"0x0000000000000000000000000000000000000000","0x0000000000000000000000000000000000000000", safe_nonce=206)
+print(safe_tx.safe_tx_hash.hex())
+```
+**Output**
+``` 
+0x34ae46cf7d884309a438a7e9a3161fa05dfc5068681ac3877a947971af845a18
+```
+
+## Offchain messages 
+Safe transaction service is able to collect the necessary offchain signatures to confirm a message  following [EIP1271](https://ethereum.org/pt/developers/tutorials/eip-1271-smart-contract-signatures/#example-eip-1271-implementation-safe).
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:68]:
```suggestion
Safe transaction service can collect the necessary offchain signatures to confirm a message using [EIP-1271](https://ethereum.org/pt/developers/tutorials/eip-1271-smart-contract-signatures/#example-eip-1271-implementation-safe).    
```

File: safe-core-api/safe-transaction-service.md [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+print(safe_tx.safe_tx_hash.hex())
+```
+**Output**
+``` 
+0x34ae46cf7d884309a438a7e9a3161fa05dfc5068681ac3877a947971af845a18
+```
+
+## Offchain messages 
+Safe transaction service is able to collect the necessary offchain signatures to confirm a message  following [EIP1271](https://ethereum.org/pt/developers/tutorials/eip-1271-smart-contract-signatures/#example-eip-1271-implementation-safe).    
+The message can be a string (EIP-191 is used to get the hash) or an object EIP-712.
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:69]:
```suggestion
The message can be a string (EIP-191 is used to get the hash) or an object EIP-712.
```

File: safe-core-api/safe-transaction-service.md [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+``` 
+0x34ae46cf7d884309a438a7e9a3161fa05dfc5068681ac3877a947971af845a18
+```
+
+## Offchain messages 
+Safe transaction service is able to collect the necessary offchain signatures to confirm a message  following [EIP1271](https://ethereum.org/pt/developers/tutorials/eip-1271-smart-contract-signatures/#example-eip-1271-implementation-safe).    
+The message can be a string (EIP-191 is used to get the hash) or an object EIP-712.    
+
+**Messages endpoints**
+- `GET /v1/safes/{address}/messages/` return the messages created for given safe address.
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:72]:
```suggestion
- `GET /v1/safes/{address}/messages/`: Return the messages created for the given Safe address.
```

File: safe-core-api/safe-transaction-service.md [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+- `POST /v1/safes/{address}/messages/` create a message with at least one signature
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:73]:
```suggestion
- `POST /v1/safes/{address}/messages/`: Create a message with at least one signature.
```

File: safe-core-api/safe-transaction-service.md [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+- `GET /v1/messages/{message_hash}/`  return a message for a given message hash
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:74]:
```suggestion
- `GET /v1/messages/{message_hash}/`: Return a message for a given message hash.
```

File: safe-core-api/safe-transaction-service.md [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+- `POST /v1/messages/{message_hash}/signatures/` add another signature to message with given message hash
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:75]:
```suggestion
- `POST /v1/messages/{message_hash}/signatures/`: Add another signature to the message with the given message hash.
```

File: safe-core-api/safe-transaction-service.md [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+
+The following sequence diagram shows an use case from a Safe shared by Alice and Bob where at least 1 signature for each one is required to fully confirm a message:
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:77]:
```suggestion
The following sequence diagram shows a use case for a Safe shared by Alice and Bob where at least one signature for each one is required to confirm a message fully:
```

File: safe-core-api/safe-transaction-service.md [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+# Create the request
+body = {
+    "signature": signature_bob.signature.hex()
+}
+requests.post(f'https://safe-transaction-goerli.safe.global/api/v1/messages/{safe_message_hash.hex()}/signatures/',json=body)
+
+```
+
+## Transactions decode
+Safe transaction service is able to decode contract interactions, to achive it, the service is periodically getting source and ABI's from different sources (Sourcify, etherscan and blockscout) using the `safe-eth-py` library.
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:140]:
```suggestion
The Safe transaction service can decode contract interactions. To achieve it, the service periodically gets source and ABIs from different sources like Sourcify, etherscan, and blockscout using the `safe-eth-py` library.   
```

File: safe-core-api/safe-transaction-service.md [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+body = {
+    "signature": signature_bob.signature.hex()
+}
+requests.post(f'https://safe-transaction-goerli.safe.global/api/v1/messages/{safe_message_hash.hex()}/signatures/',json=body)
+
+```
+
+## Transactions decode
+Safe transaction service is able to decode contract interactions, to achive it, the service is periodically getting source and ABI's from different sources (Sourcify, etherscan and blockscout) using the `safe-eth-py` library.   
+The detection of contract interactions is done in a periodic task that is executed every hour for `multisig-transaction` and `module-transactions` or every 6 hours for `multisend-transactions` on `worker-contracts-tokens`.
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:141]:
```suggestion
The detection of contract interactions is done in a periodic task executed every hour for `multisig-transaction` and `module-transactions` or every six hours for `multisend-transactions` on `worker-contracts-tokens`.
```

File: safe-core-api/safe-transaction-service.md [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+For every new contract the service try to download the source and the ABI requesting it first to `soucify`, then `etherscan` and as last chance `blockscout`. It's important to know that not all this datasources are supported or configured for every network on `safe-eth-py`.
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:142]:
```suggestion
For every new contract, the service tries to download the source, and the ABI requests it first to Sourcify, then Etherscan, and as a last chance, Blockscout. It's important to know that not all these data sources are supported or configured for every network on `safe-eth-py`.   
```

File: safe-core-api/safe-transaction-service.md [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+
+# Create the request
+body = {
+    "signature": signature_bob.signature.hex()
+}
+requests.post(f'https://safe-transaction-goerli.safe.global/api/v1/messages/{safe_message_hash.hex()}/signatures/',json=body)
+
+```
+
+## Transactions decode
```
Comment (tanay1337) [inline safe-core-api/safe-transaction-service.md:139]:
```suggestion
## Transaction Decoder
```

**Review by tanay1337 — CHANGES_REQUESTED** [review]
Thanks a lot, @moisses89!

File: safe-core-api/safe-transaction-service.md [hunk 23]
```diff
@@ -0,0 +1,185 @@
+# Safe Transaction service
+## Overview
+
+Safe transaction service keeps track of transactions sent via Safe contracts. It utilizes events (L2 chains) and tracing (L1 chains) mechanisms to index these transactions.
+
+**Key Features:**  
+
+- [**Blockchain Indexing**]() executed transactions, configuration changes, erc20/721 transfers, onchain confirmations… are automatically indexed from blockchain.
```
[the remark this answers is not in the document]
Comment (moisses89) [reply safe-core-api/safe-transaction-service.md]: I'll add blockchain indexing in other PR.

[the remark this answers is not in the document]
Comment (tanay1337) [reply safe-core-api/safe-transaction-service.md]: Gotcha, thanks! 🙌

**Review by tanay1337 — APPROVED** [review]
Thank you so much!

**Pull request merged**